### ISSUE TITLE
Make IdleStateHandler easier to extend

### ISSUE THREAD

**Issue by Spikhalskiy** [netty/netty#5616]
Netty version 4.0.40

1) IdleStateHandler has public methods like getReaderIdleTimeInMillis, but actually, this class could work with nanoseconds. It’s impossible to get precise nanoseconds setup from the entity, but why?

Maybe IdleStateHandler should expose methods like getReaderIdleTime(TimeUnit) and deprecate old versions.

2) For example, I want to implement HttpIdleStateHandler that could catch and protect from very slow clients, so if client write stream too slow - this channel would consider Idle too. To achieve this I want to put it after HttpCodec/HttpObjectAggregator and ignore a lot of phantom channelReadComplete() calls completely (For example HttpObjectAggregator could produce much more channelReadComplete than actual channelRead) and setup

```
if (readerIdleTimeNanos > 0 || allIdleTimeNanos > 0) { 
    lastReadTime = System.nanoTime(); 
    reading = false; 
}
```

directly in channelRead(), but all variables in the code above are private which makes IdleStateHandler very hard to extend.
For now, I need to copy-paste IdleStateHandler code and change this small part.

We could:
1) Make variables protected to make it possible for inherit class to access them
2) Move parts like

```
if (readerIdleTimeNanos > 0 || allIdleTimeNanos > 0) { 
    reading = true; 
    firstReaderIdleEvent = firstAllIdleEvent = true; 
}
```

```
if (readerIdleTimeNanos > 0 || allIdleTimeNanos > 0) { 
    lastReadTime = System.nanoTime(); 
    reading = false; 
}
```

to separate protected methods to make it possible to completely override channelRead and channeReadCompleted in inheritor and use them.

What do you think?

**Comment by Scottmitch** [comment]
we could add protected accessor methods and make the class more extensible if this is useful ... however we could also modify `channelReadComplete` to not set `lastReadTime` unless `channelRead` has actually been called for this  event loop cycle ... seems to make sense that the channel should only be idle if `IdleStateHandler` doesn't see a `channelRead` within the timeout period ... @normanmaurer wdyt?

**Comment by normanmaurer** [comment]
@Scottmitch makes sense. Let me do the change.

**Comment by Scottmitch** [comment]
@Spikhalskiy - now that @normanmaurer fixed  #5817 ... do you still need to extend the behavior of this handler?

**Comment by normanmaurer** [comment]
@Spikhalskiy let me close this as #5817 was fixed. Please reopen if you still need this
